/*
 *  Copyright (C) 2020 Newlogic Impact Lab Pte. Ltd.
 * 
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 * 
 *      http://www.apache.org/licenses/LICENSE-2.0
 * 
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

syntax = "proto3";

package api;
option java_package = "org.api.proto";
option java_multiple_files = true;

message KV {
  string key = 1;
  string value = 2;
}

message KVList { repeated KV vals = 1; }

message Dat {
  int32 year = 1;
  int32 month = 2;
  int32 day = 3;
}

message byteArray {
  enum Typ {
    BLOB = 0;
    AEADKEY = 1;
    ED25519PUBKEY = 2;
    ED25519PRIVKEY = 3;
    ED25519SIGNATURE = 4;
    PHOTO = 5;
  }

  Typ typ = 1;
  bytes val = 2;
}

message byteArrays { repeated byteArray vals = 1; }

message KeySet {
  bytes encryptionKey = 1;
  bytes signatureKey = 2;
  repeated byteArray verificationKeys = 3;
}

message Certificate {
  bytes privkey = 1;
  bytes pubkey = 2;
  bytes issuerkey = 3;
  bytes signature = 4;
  string desc = 5;
}

message Certificates { repeated Certificate cert = 1; }

message Ident {
  string surName = 1;
  string givenName = 2;
  string placeOfBirth = 3;
  string pin = 4;
  Dat dateOfBirth = 5;
  byteArrays photos = 6;
  bytes photo = 7;
  repeated KV privExtra = 8;
  repeated KV pubExtra = 9;
}

message Idents { repeated Ident ident = 1; }
